blaze-core/src/main/scala/org/http4s/blazecore/websocket/Http4sWSStage.scala

```diff
@@ -19,34 +19,83 @@ private[http4s] class Http4sWSStage[F[_]](ws: ws4s.Websocket[F])(
     implicit F: ConcurrentEffect[F],
     val ec: ExecutionContext)
     extends TailStage[WebSocketFrame] {
-  import Http4sWSStage.unsafeRunSync
+
+  private[this] val sentClose = new AtomicBoolean(false)
 
   def name: String = "Http4s WebSocket Stage"
 
   private val deadSignal: Signal[F, Boolean] =
-    unsafeRunSync[F, Signal[F, Boolean]](async.signalOf[F, Boolean](false))
+    F.toIO(async.signalOf[F, Boolean](false)).unsafeRunSync()
 
   //////////////////////// Source and Sink generators ////////////////////////
 
   def snk: Sink[F, WebSocketFrame] = _.evalMap { frame =>
+    F.delay(sentClose.get()).flatMap { closeSent =>
+      if (!closeSent) {
+        frame match {
+          case c: Close =>
+            F.delay(sentClose.compareAndSet(false, true)).flatMap {
+              //write the close frame, as it was set atomically
+              case true =>
+                writeFrame(c)
+              case false =>
+                //Close was set concurrently, so do nothing
+                F.unit
+            }
+          case _ =>
+            writeFrame(frame)
+        }
+      } else {
+        //Close frame has been sent. Send no further data
+        F.unit
+      }
+    }
+  }
+
+  private[this] def writeFrame(frame: WebSocketFrame): F[Unit] =
     F.async[Unit] { cb =>
       channelWrite(frame).onComplete {
         case Success(res) => cb(Right(res))
         case Failure(t @ Command.EOF) => cb(Left(t))
         case Failure(t) => cb(Left(t))
       }(directec)
     }
-  }
 
+  /** The websocket input stream
+    *
+    * Note: On receiving a close, we MUST send a close back, as stated in section
+    * 5.5.1 of the websocket spec: https://tools.ietf.org/html/rfc6455#section-5.5.1
+    *
+    * @return
+    */
   def inputstream: Stream[F, WebSocketFrame] = {
     val t = F.async[WebSocketFrame] { cb =>
       def go(): Unit =
         channelRead().onComplete {
           case Success(ws) =>
             ws match {
-              case c @ Close(_) =>
-                unsafeRunSync[F, Unit](deadSignal.set(true))
-                cb(Right(c)) // With Dead Signal Set, Return callback with the Close Frame
+              case c: Close =>
+                if (sentClose.get()) {
+                  //We sent the close signal, thus we can be sure to end the stream here.
+                  F.toIO(deadSignal.set(true)).unsafeRunSync()
```

Comment:
These were done as unsafeRunSync calls, due to issues with tests.

Originally, they were

```scala
F.toIO(deadSignal.set(true)).unsafeRunAsync(_ => cb(Right(c))
```